qiling/core.py

```diff
@@ -685,6 +687,22 @@ def _exit_trap(ql):
 
             self.hook_address(_exit_trap, self._exit_trap_addr)
 
+        # Stop when running to an unmapped IP
+        if self.stop_options.unmapped_ip:
+            def _check_ip(ql, access, address, size, value):
+                # make sure we got here for the right reason
+                if access == uc.UC_MEM_FETCH_UNMAPPED:
+                    self.log.info(f'Process returned from diverted execution flow to 0x{address:x}')
+                    # to gracefully recover from memory read/write error, you have to memory map the invalid page before you return true
+                    # https://github.com/unicorn-engine/unicorn/issues/1484#issuecomment-970727708
+                    self.mem.map(address//0x1000*0x1000, 0x1000, info='[BufferOverflow guard]')
```

Comment:
Suggesting to use `mem.align` and `mem.pagesize` to maintain compatibility.

Reply:
I was trying using `ql.mem.align` but it end up allocating memory after the address pointed in the instruction pointer:

```
[=]     Process returned from diverted execution flow to 0x61616168
[=]     Allocated aligned memory at 0x61617000
```

```python
self.log.info(f'Process returned from diverted execution flow to 0x{address:x}')
addr = self.mem.align(address, 0x1000)
self.log.info(f'Allocated aligned memory at 0x{addr:x}')
self.mem.map(addr, 0x1000, info='[BufferOverflow guard]')
```

And thus Unicorn complains

Reply:
You're right, `mem.align` rounds the address up.. What about this one:
`self.mem.map(address & ~(self.mem.pagesize - 1), self.mem.pagesize, info='[guard]')`

Reply:
maybe we can add an optional parameter to `mem.align` in order to specify if the address we are passing should be inside the aligned memory and thus doing floor rounding. This could be used often.

I like your proposal for the rounding

Reply:
I don't think letting an optional parameter decide the rounding direction is a good idea.. an optional parameter may be used to slightly change the behavior of the function, not change it entirely.

Normally an `align` method would mean aligning the value by roudning it **down**, so I think it may be best to change the implementation in the `mem` module to have it rounding down one would expect. It looks like there are only a few usages of the `mem.align` currently, so it will be easy to revisit its usages and provide them with an alternate method of rounding up.

If it's OK with you, let's leave the `mem.align` method intact for now, and once I fix its functionality (to round down instead of up), we'll use it as intended. What do you think?

Reply:
Make sense, I will exclude the `mem.align` commit from the PR